### DISCUSSION TITLE
SHAssocEnumHandlers crashes spectacularly on Windows 7

### DISCUSSION THREAD

**Discussion by AliveDevil** [microsoft/CsWin32#530] [Q&A] [answered]
Is there anything I can do, targeting .NET Framework 4.7.2+, that prevents .NET Framework from crashing upon trying to do anything with `System.StubHelpers.InterfaceMarshaler.ConvertToManaged`?

```
Unbehandelte Ausnahme: System.Runtime.InteropServices.COMException: Unbekannter Fehler (Ausnahme von HRESULT: 0x80004005 (E_FAIL))
   bei System.StubHelpers.InterfaceMarshaler.ConvertToManaged(IntPtr pUnk, IntPtr itfMT, IntPtr classMT, Int32 flags)
   bei Windows.Win32.PInvoke.SHAssocEnumHandlers(PCWSTR pszExtra, ASSOC_FILTER afFilter, IEnumAssocHandlers& ppEnumHandler)
   bei Windows.Win32.PInvoke.SHAssocEnumHandlers(String pszExtra, ASSOC_FILTER afFilter, IEnumAssocHandlers& ppEnumHandler) in C:\Users\jmalek\Desktop\assoc\assoc\Microsoft.Windows.CsWin32\Microsoft.Windows.CsWin32.SourceGenerator\Windows.Win32.PInvoke.Shell32.g.cs:Zeile 182.
   bei Program.Main() in C:\Users\jmalek\Desktop\assoc\assoc\Program.cs:Zeile 93.
```

Exception above happens immediatly with
```
[STAThread]
private static void Main()
{
    SHAssocEnumHandlers(".mp4", ASSOC_FILTER.ASSOC_FILTER_RECOMMENDED, out var handlers).ThrowOnFailure();
}
```

Setting `allowMarshaling` to `false` mitigates this issue, but is rather cumbersome to work with.

**Comment by AArnott** [comment]
It doesn't crash for me. Here is my project:
[CsWin32Sandbox.zip](https://github.com/microsoft/CsWin32/files/8636919/CsWin32Sandbox.zip)

**Reply by AliveDevil** [reply]
Interesting. I’ll try and get my VM set up again to check.

**Reply by AliveDevil** [reply]
So … `dotnet build -c Release -f net472`, zipped, sent to Windows 7, extracted, and it crashes.

```
.\CsWin32Sandbox.exe

Unbehandelte Ausnahme: System.Runtime.InteropServices.COMException: Unbekannter Fehler (Ausnahme von HRESULT: 0x80004005 (E_FAIL))
   bei System.StubHelpers.InterfaceMarshaler.ConvertToManaged(IntPtr pUnk, IntPtr itfMT, IntPtr classMT, Int32 flags)
   bei Windows.Win32.PInvoke.SHAssocEnumHandlers(PCWSTR pszExtra, ASSOC_FILTER afFilter, IEnumAssocHandlers& ppEnumHandler)
   bei Program.<Main>$(String[] args) in C:\Users\jmalek\Desktop\CsWin32Sandbox\CsWin32Sandbox\Program.cs:Zeile 6.
```

So … according to the observations here it is based on locale of the installed system - my installation is on de_DE.

**Reply by AArnott** [reply]
I wouldn't expect the behavior of this to vary by locale. I wonder if it works on en-US Windows 7. Could it just be a Windows 7 thing?

**Reply by AliveDevil** [reply]
Going to check. Used some Windows 7 AIO-ISO which supposedly includes updates until April, though the .NET Framework updates fail as well. Checking both Windows 7 en-US and de-DE from original ISO now.
I thought you checked with Windows 7 as well, on Windows 10 and 11 it doesn't crash here as well.

**Reply by AArnott** [reply]
No, I never test on Windows 7, which [hasn't been supported since 2020](https://www.microsoft.com/en-US/windows/windows-7-end-of-life-support-information?msclkid=e49abbcccfa211ec83361655dd01a7ac).

**Reply by AliveDevil** [reply]
Fair enough. We moved forward dropping support for anything older than Server 2016.
Thought someone could shed some light on this, as the earliest I've found this issue mentioned was in [2018](https://stackoverflow.com/q/48400038) (where it still was supported).

**Reply by AliveDevil** [reply]
It is indeed a general Windows 7 issue. On de-DE and en-US installed from the original SP1 installer ISO updated through Windows Update results in the very same error being created.

**Comment by AliveDevil** [comment, marked as answer]
Originally first discovered in 2018[^1] this is an issue in Windows 7, which hasn't been adressed until its deprecation.
Resolution: Either disallow marshalling (using vtable-ptrs), or drop Windows 7 support.

[^1]: https://stackoverflow.com/q/48400038
